Shell/Builtin.cpp

```diff
@@ -0,0 +1,743 @@
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    String new_path;
+
+    if (!arg_path) {
+        new_path = home;
+        if (cd_history.is_empty() || cd_history.last() != home)
+            cd_history.enqueue(home);
+    } else {
+        if (cd_history.is_empty() || cd_history.last() != arg_path)
+            cd_history.enqueue(arg_path);
+        if (strcmp(argv[1], "-") == 0) {
+            char* oldpwd = getenv("OLDPWD");
+            if (oldpwd == nullptr)
+                return 1;
+            new_path = oldpwd;
+        } else if (arg_path[0] == '/') {
+            new_path = argv[1];
+        } else {
+            StringBuilder builder;
+            builder.append(cwd);
+            builder.append('/');
+            builder.append(arg_path);
+            new_path = builder.to_string();
+        }
+    }
+
+    LexicalPath lexical_path(new_path);
+    if (!lexical_path.is_valid()) {
+        printf("LexicalPath failed to canonicalize '%s'\n", new_path.characters());
+        return 1;
+    }
+    const char* path = lexical_path.string().characters();
+
+    struct stat st;
+    int rc = stat(path, &st);
+    if (rc < 0) {
+        printf("stat(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    if (!S_ISDIR(st.st_mode)) {
+        printf("Not a directory: %s\n", path);
+        return 1;
+    }
+    rc = chdir(path);
+    if (rc < 0) {
+        printf("chdir(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    setenv("OLDPWD", cwd.characters(), 1);
+    cwd = lexical_path.string();
+    setenv("PWD", cwd.characters(), 1);
+    return 0;
+}
+
+int Shell::builtin_cdh(int argc, const char** argv)
+{
+    int index = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(index, "Index of the cd history entry (leave out for a list)", "index", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (index == -1) {
+        if (cd_history.size() == 0) {
+            printf("cdh: no history available\n");
+            return 0;
+        }
+
+        for (int i = cd_history.size() - 1; i >= 0; --i)
+            printf("%lu: %s\n", cd_history.size() - i, cd_history.at(i).characters());
+        return 0;
+    }
+
+    if (index < 1 || (size_t)index > cd_history.size()) {
+        fprintf(stderr, "cdh: history index out of bounds: %d not in (0, %zu)\n", index, cd_history.size());
+        return 1;
+    }
+
+    const char* path = cd_history.at(cd_history.size() - index).characters();
+    const char* cd_args[] = { "cd", path };
+    return Shell::builtin_cd(2, cd_args);
+}
+
+int Shell::builtin_dirs(int argc, const char** argv)
+{
+    // The first directory in the stack is ALWAYS the current directory
+    directory_stack.at(0) = cwd.characters();
+
+    if (argc == 1) {
```

Comment:
It's almost always wrong to check `argc` against something to learn whether some argument is specified or not. Better delay it until `args_parser` parses the list of `paths` and look at that.

Reply:
It's checking against `1` though, what else could be there other than the program name?

Reply:
Made there are some options that modify how the output is formatted but not what it should be. Perhaps there aren't any such options now, but still.

Reply:
o...kay...

Reply:
from `man zshbuiltins`:
```
       dirs [ -c ] [ arg ... ]
       dirs [ -lpv ]
              With  no arguments, print the contents of the directory stack.  Direc‐
              tories are added to this stack with the  pushd  command,  and  removed
              with  the  cd or popd commands.  If arguments are specified, load them
              onto the directory stack, replacing anything that was there, and  push
              the current directory onto the stack.

              -c     clear the directory stack.

              -l     print  directory  names in full instead of using of using ~ ex‐
                     pressions (see Dynamic and Static  named  directories  in  zsh‐
                     expn(1)).

              -p     print directory entries one per line.

              -v     number the directories in the stack when printing.
```

"With no arguments, print the contents of the directory stack", I'm just gonna say this is to-spec and leave it be.
we don't really have a reasonable way to detect empty arguments other than `argc == 1`.

Reply:
I don't really insist, so feel free to leave it as is, but let me still rebut your argument just for the sake of it :wink:

> from `man zshbuiltins`
> I'm just gonna say this is to-spec and leave it be

Heh, since when is the manpage for zshbuiltins a spec for our Shell?

Here's a simple explanation of the difference between a man page and a spec: a man page is descriptive — if what's written in the man page differs from the actual behavior, the man page is to be fixed, while the spec is prescriptive — an implementation is to be fixed if its behavior differs from what the spec mandates. That is, of course, if the implementation even wishes to follow the spec (like say LibWeb aims to follow the HTML spec, and our Unix layer mostly aims to follow POSIX where it's reasonable).

> we don't really have a reasonable way to detect empty arguments other than `argc == 1`.

Sure we do. While you think of it as of "empty arguments" (and it's even written that way in the zsh man page you cited), the actual thing to check for is that no *directory* is specified in argv. Like I said before, there may exist options that modify the output, but do not cause `dirs` to fall off the "no arguments ⇒ print directory stack" case. In fact, the `-lpv` options listed in the man page you cite are exactly those options: `dirs -l` works the same as just `dirs`, but formats output a bit differently, — and `argc` would be 2 in this case.

Some what the man page should *actually* say is that it lists the directories in the stack if neither any *paths* nor the `-c` option is specified. The `-lpv` options do not cause it to choose the second behavior, only to format the output differently.

If you need any further proof, here's zsh's `dirs` with argc equal to 1 and 4, behaving the same:

```
% dirs
/tmp ~
% dirs -l -p -v
0	/tmp
1	/home/sergey
```

So, what does this all mean for our case? It's simple: we should check that `paths.is_empty() && !clear`, and list the paths in that case; otherwise push new paths to the stack.

Reply:
> since when is the manpage for zshbuiltins a spec for our Shell

ouch.

I see your point.
Then we don't really need the check or its branch at all, the code below should be sufficient with your addition.

and this reminds me that we still don't have a manpage for the shell or its builtins, someone should get on that sooner rather than later...